File: CalculoSEF95.m
function SpectralEdgeFrequency95 = CalculoSEF95(PSD, f, banda)
% CALCULOMF Calcula la frecuencia mediana de la distribucion
%       contenida en PSD.
%
%		SPECTRALEDGEFREQUENCY95 = CALCULOSEF95(PSD, F, BANDA), calcula la frecuencia 
%       espectral límite al 95% de la distribucion PSD, indexada por el vector F, entre las 
%       frecuencias indicadas en BANDA.
%
%       En SPECTRALEDGEFREQUENCY95 se devuelve la frecuencia espectral límite al 95% calculada.
%

%
% Versión: 2.0
%
% Fecha de creación: 21 de Abril de 2021
%
% Autor: Víctor Gutiérrez de Pablo (a partir de un script de Jesús Poza Crespo)
%

% Se inicializa la variable de salida.
SpectralEdgeFrequency95 = [];

% Se buscan los índices positivos en la banda de paso considerada.
indbanda = find((f >= banda(1)) & (f <= banda(2)));
% Potencia total para el espectro positivo
potenciatotal = sum(PSD(indbanda));
% Se suman los valores de potencia relativa para el espectro positivo
vectorsuma = cumsum(PSD(indbanda));

% Se coge el índice para el cual se tiene el 95% de la potencia total.
ind95 = max(find(vectorsuma <= (0.95*potenciatotal)));
indmedia95 = indbanda(ind95);
% Si no se ha seleccionado ningún índice es porque en el primer valor esta
% mas del 95% de la potencia total
if isempty(indmedia95),
    indmedia95 = indbanda(1);
end % Fin del 'if' que comprueba si hay algun iíndice

% Se toma la frecuencia con la potencia al 95% (frecuencia al 95% de la potencia)
SpectralEdgeFrequency95 = f(indmedia95);

clear PSD f banda indbanda potencia toal vectorsuma indmitad indmedia

File: CalculoSEF95.py
import numpy as np

def calcular_sef95(psd: np.ndarray, f: np.ndarray, banda: list[float]) -> float | None:
    """
    Calcula la Frecuencia Espectral Límite al 95% (SEF95) de la distribución 
    de densidad espectral de potencia (PSD) dentro de una banda de frecuencia específica.

    Args:
        psd: Array de NumPy que representa la Densidad Espectral de Potencia.
        f: Array de NumPy que representa el vector de frecuencias correspondiente a psd.
        banda: Lista o tupla con dos elementos [f_min, f_max] especificando 
               la banda de frecuencia de interés.

    Returns:
        La frecuencia espectral límite al 95% calculada dentro de la banda especificada.
        Devuelve None si no hay datos en la banda, si la potencia total es cero o negativa,
        o si ocurre un error.
    """

    if len(psd) != len(f):
        raise ValueError("psd y f deben tener la misma longitud.")
    if not isinstance(banda, (list, tuple)) or len(banda) != 2:
        raise ValueError("banda debe ser una lista o tupla de dos elementos [f_min, f_max].")
    if banda[0] > banda[1]:
        raise ValueError("El primer elemento de banda (f_min) no puede ser mayor que el segundo (f_max).")

    # Encontrar índices dentro de la banda de frecuencia
    indbanda = np.where((f >= banda[0]) & (f <= banda[1]))[0]

    if indbanda.size == 0:
        print(f"Advertencia: No se encontraron frecuencias en la banda especificada [{banda[0]}, {banda[1]}].")
        return None

    psd_banda = psd[indbanda]
    f_banda = f[indbanda]

    # Potencia total en la banda
    potencia_total = np.sum(psd_banda)

    if potencia_total <= 0:
        print(f"Advertencia: La potencia total en la banda [{banda[0]}, {banda[1]}] es cero o negativa.")
        return None

    # Suma acumulada de la potencia en la banda
    vector_suma = np.cumsum(psd_banda)

    # Encontrar el índice donde la suma acumulada alcanza el 95% de la potencia total
    indices_95 = np.where(vector_suma <= (0.95 * potencia_total))[0]

    if indices_95.size == 0:
        # Si ningún índice cumple (es decir, vector_suma[0] > 0.95 * potencia_total),
        # la SEF95 es la primera frecuencia de la banda.
        ind_sef95_en_banda = 0
    else:
        # El índice deseado es el último que cumple la condición (equivalente a max(find(...)))
        ind_sef95_en_banda = indices_95[-1]

    # La SEF95 es la frecuencia correspondiente a ese índice
    spectral_edge_frequency_95 = f_banda[ind_sef95_en_banda]

    return float(spectral_edge_frequency_95) 